src/plugins/shiftmanager/shiftmanager.go

```diff
@@ -0,0 +1,267 @@
+		RadonURL:               shiftInfo.RadonURL,
+		Checksum:               shiftInfo.Checksum,
+		WaitTimeBeforeChecksum: shiftInfo.WaitTimeBeforeChecksum,
+	}
+
+	switch typ {
+	case ShiftTypeReshard:
+		cfg.ToFlavor = shift.ToRadonDBFlavor
+	case ShiftTypeRebalance:
+		cfg.ToFlavor = shift.ToMySQLFlavor
+	default:
+		shiftMgr.log.Error(`shift.wrong.shifttype: [%+v].`, typ)
+		return nil, fmt.Errorf("shift.wrong.shifttype: [%+v].", typ)
+	}
+
+	shiftMgr.log.Info("shift.cfg:%+v", cfg)
+
+	slog := sxlog.NewStdLog(sxlog.Level(sxlog.WARNING))
+	return shift.NewShift(slog, cfg), nil
+}
+
+// StartAnInstance -- used to start a new shift instance
+func (shiftMgr *ShiftManager) StartShiftInstance(key string, shift shift.ShiftHandler, typ ShiftType) error {
+	shiftMgr.mu.Lock()
+	defer shiftMgr.mu.Unlock()
+	// check if the instance specified by key has been already in instancesAlived
+	if _, ok := shiftMgr.instancesAlived[key]; ok {
+		return fmt.Errorf("shift.instance[%v].is.already.running", key)
+	}
+
+	if typ == ShiftTypeNone {
+		return fmt.Errorf("shift.instance.type.should.not.be.none")
+	}
+	if err := shift.Start(); err != nil {
+		shiftMgr.log.Error("shift.instance.start.error:%+v", err)
+		return err
+	}
+
+	if err := shiftMgr.addAnAlivedInstance(key, typ, shift); err != nil {
+		shiftMgr.log.Error("shiftMgr.add.instance.error:%+v", err)
+		return err
+	}
+
+	return nil
+}
+
+// addAnInstance -- used to add an shift instance to shift manager
+// key: for reshard, key is db_table, for rebalance, key is db_table_backend
+func (shiftMgr *ShiftManager) addAnAlivedInstance(key string, typ ShiftType, shift shift.ShiftHandler) error {
+	if len(shiftMgr.instancesAlived) < maxShiftInstancesAlived {
+		shiftMgr.instancesAlived[key] = &shiftInstancesAlived{
+			shift:     shift,
+			status:    ShiftStatusMigrating,
+			progress:  "",
+			shiftType: typ,
+		}
+		return nil
+	}
+	return fmt.Errorf("shift.instances.num.exceeding.10.limits")
+}
+
+func (shiftMgr *ShiftManager) updateFinishedInstance(key string, status ShiftStatus, typ ShiftType) {
+	shiftMgr.mu.Lock()
+	defer shiftMgr.mu.Unlock()
+	finished := &shiftInstancesFinished{
+		status:    status,
+		progress:  "",
+		shiftType: typ,
+	}
+	// the finished instance added into instancesFinished
+	shiftMgr.instancesFinished[key] = finished
+	// the finished instance in instancesAlived should be removed
+	delete(shiftMgr.instancesAlived, key)
+}
+
+// WaitInstanceFinishThread -- used to start a thread to excute wait finish in background.
+func (shiftMgr *ShiftManager) WaitInstanceFinishThread(key string) error {
+	shiftMgr.mu.Lock()
+	defer shiftMgr.mu.Unlock()
+	// get instance specified by key
+	instance, ok := shiftMgr.instancesAlived[key]
+	if !ok {
+		shiftMgr.log.Error("shift.manager.wait.thread.start.error:instance[%v].not.found", key)
+		return fmt.Errorf("shift.manager.wait.thread.start.error:instance[%v].not.found", key)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
```

Comment:
`rebalance` need wait the `shift` finished in a instance.

Reply:
I add a WaitInstanceFinish(key string) error function to meet you demands.

Reply:
I add a WaitInstanceFinish(key string) error function to meet you demands.